danielcondemarin/serverless-nextjs-plugin
# Add better support for static assets

#### What does this PR do?

This pull request addresses feature request #89 and issue #88 by adding support for static file serving for

* static files specified in a static folder
  * The code will check to see if any files exist within a static folder. If so, it will create a resource in API Gateway to proxy all requests from http://mydomain.com/static to the static folder location in S3.
* public files specified in a public folder
  * The code will check to see if any files exist within a public folder. If so, it will create a resource in API Gateway for each file and proxy all requests from http://mydomain.com/mypublicfile.xyz to the public file location in S3.
* NextJS static assets
  * The code will create a resource in API Gateway to proxy all requests from http://mydomain.com/_next to the _next folder location in S3.

This also removes the previous functionality of providing `routes` in serverless.yml.

#### What are the relevant tickets?

* #89 
* #88

#### Where should the reviewer start?

The bulk of the code change is in addCustomStackResources.js

#### How should this be manually tested?

Obviously run the updated unit tests and ensure I've covered all cases.

You'll then want to run this new code to deploy some AWS infrastructure. Without changing anything, you should be able to see a few extra resources in your API Gateway dashboard.

Specifically, you'll want to see an entry for

* `_next/{proxy+}`
* `static/{proxy+}`
* Each public file, i.e. `/manifest.json`

Then ensure you can browse each file once deployed, such as

* https://mydomain.com/_next/static/runtime/main-29fa3c7761b55170bf85.js
* https://mydomain.com/static/manifest.json
* https://mydomain.com/favicon.ico

<img width="1032" alt="Screenshot 2019-06-20 at 14 10 05" src="https://user-images.githubusercontent.com/2536442/59851774-60693b80-9365-11e9-9276-2ac030875cfe.png">

Also your `<head>` element should only contain relative paths for assets (unless `assetPrefix` is specified in next.config.js).

<img width="1245" alt="Screenshot 2019-06-20 at 14 16 03" src="https://user-images.githubusercontent.com/2536442/59852099-0b79f500-9366-11e9-95ad-0853d9a5a262.png">
